cloudsync/cs.py

```diff
@@ -260,24 +260,46 @@ def start(self, *, daemon=True, **kwargs):
         self.nmgr.notify(Notification(SourceEnum.SYNC, NotificationType.STARTED, None))
         self.nmgr.start(**kwargs)
 
-
     def stop(self, forever=True, wait=True):
         """
         Stops the cloudsync service.
 
         Args:
             forever: If false is passed, then handles are left open for a future start.  Generally used for tests only.
+            wait: If false, notification manager thread is signaled to stop, but the caller does not wait for it.
+
+        NOTE: The wait parameter does not affect the sync manager or event manager threads. These threads need to write
+        out state on CloudSync shutdown, and as such we wait on them regardless of the wait parameter value.
+        """
+        self._stop(forever=forever, wait=wait, join=True)
+
+    @staticmethod
+    def stop_all(syncs: List["CloudSync"]):
         """
+        Convenience function for stopping multiple CloudSyncs efficiently.
+        """
+        for cs in syncs:
+            cs._stop(forever=True, wait=False, join=False)  # pylint: disable=protected-access
+        for cs in syncs:
+            cs._join()  # pylint: disable=protected-access
+
+    def _stop(self, forever=True, wait=True, join=True):
         log.info("stopping sync: %s", self.storage_label())
         self.smgr.stop(forever=forever, wait=wait)
         self.emgrs[0].stop(forever=forever, wait=wait)
         self.emgrs[1].stop(forever=forever, wait=wait)
         self.nmgr.stop(forever=forever, wait=wait)
+        if join:
+            self._join()
+
+    def _join(self):
         if self.sthread:
             self.sthread.join()
             self.ethreads[0].join()
             self.ethreads[1].join()
-            self.nmgr.wait()  # TODO: wait() above instead of join()?
+            # TODO: Fix this. Calling wait() after stop() does nothing for Runnable-managed threads.
```

Comment:
Once you change stop(), this comment can go away

Reply:
removed